File: research/astronet/light_curve_util/median_filter.py
# Copyright 2018 The TensorFlow Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Utility function for smoothing data using a median filter."""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np


def median_filter(x, y, num_bins, bin_width=None, x_min=None, x_max=None):
  """Computes the median y-value in uniform intervals (bins) along the x-axis.

  The interval [x_min, x_max) is divided into num_bins uniformly spaced
  intervals of width bin_width. The value computed for each bin is the median
  of all y-values whose corresponding x-value is in the interval.

  NOTE: x must be sorted in ascending order or the results will be incorrect.

  Args:
    x: 1D array of x-coordinates sorted in ascending order. Must have at least 2
        elements, and all elements cannot be the same value.
    y: 1D array of y-coordinates with the same size as x.
    num_bins: The number of intervals to divide the x-axis into. Must be at
        least 2.
    bin_width: The width of each bin on the x-axis. Must be positive, and less
        than x_max - x_min. Defaults to (x_max - x_min) / num_bins.
    x_min: The inclusive leftmost value to consider on the x-axis. Must be less
        than or equal to the largest value of x. Defaults to min(x).
    x_max: The exclusive rightmost value to consider on the x-axis. Must be
        greater than x_min. Defaults to max(x).

  Returns:
    1D NumPy array of size num_bins containing the median y-values of uniformly
    spaced bins on the x-axis.

  Raises:
    ValueError: If an argument has an inappropriate value.
  """
  if num_bins < 2:
    raise ValueError("num_bins must be at least 2. Got: %d" % num_bins)

  # Validate the lengths of x and y.
  x_len = len(x)
  if x_len < 2:
    raise ValueError("len(x) must be at least 2. Got: %s" % x_len)
  if x_len != len(y):
    raise ValueError("len(x) (got: %d) must equal len(y) (got: %d)" % (x_len,
                                                                       len(y)))

  # Validate x_min and x_max.
  x_min = x_min if x_min is not None else x[0]
  x_max = x_max if x_max is not None else x[-1]
  if x_min >= x_max:
    raise ValueError("x_min (got: %d) must be less than x_max (got: %d)" %
                     (x_min, x_max))
  if x_min > x[-1]:
    raise ValueError(
        "x_min (got: %d) must be less than or equal to the largest value of x "
        "(got: %d)" % (x_min, x[-1]))

  # Validate bin_width.
  bin_width = bin_width if bin_width is not None else (x_max - x_min) / num_bins
  if bin_width <= 0:
    raise ValueError("bin_width must be positive. Got: %d" % bin_width)
  if bin_width >= x_max - x_min:
    raise ValueError(
        "bin_width (got: %d) must be less than x_max - x_min (got: %d)" %
        (bin_width, x_max - x_min))

  bin_spacing = (x_max - x_min - bin_width) / (num_bins - 1)

  # Bins with no y-values will fall back to the global median.
  result = np.repeat(np.median(y), num_bins)

  # Find the first element of x >= x_min. This loop is guaranteed to produce
  # a valid index because we know that x_min <= x[-1].
  x_start = 0
  while x[x_start] < x_min:
    x_start += 1

  # The bin at index i is the median of all elements y[j] such that
  # bin_min <= x[j] < bin_max, where bin_min and bin_max are the endpoints of
  # bin i.
  bin_min = x_min  # Left endpoint of the current bin.
  bin_max = x_min + bin_width  # Right endpoint of the current bin.
  j_start = x_start  # Inclusive left index of the current bin.
  j_end = x_start  # Exclusive end index of the current bin.

  for i in range(num_bins):
    # Move j_start to the first index of x >= bin_min.
    while j_start < x_len and x[j_start] < bin_min:
      j_start += 1

    # Move j_end to the first index of x >= bin_max (exclusive end index).
    while j_end < x_len and x[j_end] < bin_max:
      j_end += 1

    if j_end > j_start:
      # Compute and insert the median bin value.
      result[i] = np.median(y[j_start:j_end])

    # Advance the bin.
    bin_min += bin_spacing
    bin_max += bin_spacing

  return result

File: research/astronet/light_curve_util/cc/median_filter.cc
/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "light_curve_util/cc/median_filter.h"

#include "absl/strings/substitute.h"
#include "light_curve_util/cc/median.h"

using absl::Substitute;
using std::min;
using std::vector;

namespace astronet {

bool MedianFilter(const vector<double>& x, const vector<double>& y,
                  int num_bins, double bin_width, double x_min, double x_max,
                  vector<double>* result, std::string* error) {
  const std::size_t x_size = x.size();
  if (x_size < 2) {
    *error = Substitute("x.size() must be greater than 1. Got: $0", x_size);
    return false;
  }
  if (x_size != y.size()) {
    *error = Substitute("x.size() (got: $0) must equal y.size() (got: $1)",
                        x_size, y.size());
    return false;
  }
  const double x_first = x[0];
  const double x_last = x[x_size - 1];
  if (x_first >= x_last) {
    *error = Substitute(
        "The first element of x (got: $0) must be less than the last "
        "element (got: $1). Either x is not sorted or all elements are "
        "equal.",
        x_first, x_last);
    return false;
  }
  if (x_min >= x_max) {
    *error = Substitute("x_min (got: $0) must be less than x_max (got: $1)",
                        x_min, x_max);
    return false;
  }
  if (x_min > x_last) {
    *error = Substitute(
        "x_min (got: $0) must be less than or equal to the largest value of x "
        "(got: $1)",
        x_min, x_last);
    return false;
  }
  if (bin_width <= 0) {
    *error = Substitute("bin_width must be positive. Got: $0", bin_width);
    return false;
  }
  if (bin_width >= x_max - x_min) {
    *error = Substitute(
        "bin_width (got: $0) must be less than x_max - x_min (got: $1)",
        bin_width, x_max - x_min);
    return false;
  }
  if (num_bins < 2) {
    *error = Substitute("num_bins must be greater than 1. Got: $0", num_bins);
    return false;
  }

  result->resize(num_bins);

  // Compute the spacing between midpoints of adjacent bins.
  double bin_spacing = (x_max - x_min - bin_width) / (num_bins - 1);

  // Create a vector to hold the values of the current bin on each iteration.
  // Its initial size is twice the expected number of points per bin if x
  // values are uniformly spaced. It will be expanded as necessary.
  int points_per_bin =
      1 + static_cast<int>(x_size * min(1.0, bin_width / (x_last - x_first)));
  vector<double> bin_values(2 * points_per_bin);

  // Create a vector to hold the indices of any empty bins.
  vector<int> empty_bins;

  // Find the first element of x >= x_min. This loop is guaranteed to produce
  // a valid index because we know that x_min <= x_last.
  int x_start = 0;
  while (x[x_start] < x_min) ++x_start;

  // The bin at index i is the median of all elements y[j] such that
  // bin_min <= x[j] < bin_max, where bin_min and bin_max are the endpoints of
  // bin i.
  double bin_min = x_min;              // Left endpoint of the current bin.
  double bin_max = x_min + bin_width;  // Right endpoint of the current bin.
  int j_start = x_start;  // Index of the first element in the current bin.
  int j = x_start;        // Index of the current element in the current bin.

  for (int i = 0; i < num_bins; ++i) {
    // Move j_start to the first index of x >= bin_min.
    while (j_start < x_size && x[j_start] < bin_min) ++j_start;

    // Accumulate values y[j] such that bin_min <= x[j] < bin_max. After this
    // loop, j is the exclusive end index of the current bin.
    j = j_start;
    while (j < x_size && x[j] < bin_max) {
      if (j - j_start >= bin_values.size()) {
        bin_values.resize(2 * bin_values.size());  // Expand if necessary.
      }
      bin_values[j - j_start] = y[j];
      ++j;
    }

    int n = j - j_start;  // Number of points in the bin.
    if (n == 0) {
      empty_bins.push_back(i);  // Empty bin.
    } else {
      // Compute and insert the median bin value.
      (*result)[i] = InPlaceMedian(bin_values.begin(), bin_values.begin() + n);
    }

    // Advance the bin.
    bin_min += bin_spacing;
    bin_max += bin_spacing;
  }

  // For empty bins, fall back to the median y value between x_min and x_max.
  if (!empty_bins.empty()) {
    double median = Median(y.begin() + x_start, y.begin() + j);
    for (int i : empty_bins) {
      (*result)[i] = median;
    }
  }
  return true;
}

}  // namespace astronet
